Feature/phpunit6 support

omissis opened the pull request:
I updated the tests, dependencies and travis config to support phpunit 6.

omissis committed:
Update travis configuration to php7 only, while adding a few optimizations.

theofidry commented on .travis.yml:
```diff
@@ -1,24 +1,28 @@
-  - 5.4
-  - hhvm
-  - hhvm-nightly
+    - '7.0'
```
we're still supporting PHP 5.6 AFAIK :/

omissis replied:
See comment below

theofidry commented on .travis.yml:
```diff
@@ -1,24 +1,28 @@
+    - '7.1'
+    - nightly
+    - hhvm
+    - hhvm-nightly
```
we can skip `hhvm-nightly`

omissis replied:
👍

theofidry commented on composer.json:
```diff
@@ -12,15 +12,15 @@
     "require": {
-        "php": ">=5.4.0",
-        "phpunit/phpunit": "^4.0|^5.0"
+        "php": "^7.0",
```
we allow PHP 5.6, same for PHPUnit, 5.x should be allowed ass well

omissis replied:
uhm ok so should this support both phpunit 5 and phpunit 6? I thought the bump would've caused a new major to be released, but I will check if it's doable to support both at the same time :)

theofidry commented on tests/Humbug/Test/Phpunit/Logger/JsonLoggerTest.php:
```diff
@@ -1,31 +1,37 @@
+use Humbug\Phpunit\Logger\JsonLogger;
+use LogicException;
+
+class JsonLoggerTest extends \PHPUnit\Framework\TestCase
```
you can simplify the FQCN by importing it

omissis replied:
👍

theofidry reviewed (changes requested):
Thanks for the PR! Left some comments

padraic commented:
We likely need to support both PHPUnit 5 and 6, but I assume composer can resolve the versions? We will effectively a 1.x here for 5, and a 2.x for 6. I'd prefer to avoid any kind of class-alias shim if possible.

So keep the PR live - once we bump 1.0 to a maintenance branch, we can probably then merge it.

theofidry commented:
@padraic unless I'm missing something we shouldn't need any shim here to support both

omissis commented:
Afaik it should be possible to support both thanks to phpunit 5 being forward compatible (at least for a number of classes, including TestCase), but I thought that a new major with php 7 and phpunit 6 was the preferable option. no problem either way :)

padraic commented:
Also, I'd wait until the PR can rebase to #5 which might explain the shim comment - the code specifically extends from non-namespaced PHPUnit classes. Not sure if that can be worked around short of version bumping.

padraic commented:
@theofidry Are we considering #5 relatively benign to merge, or wait for someone (other than me) from @humbug/core?

theofidry commented:
@padraic I wouldn't say it's trivial but I think it's ok for this one :)

padraic commented:
@omissis I've merged the other PR which adds additional tests, if you could rebase and include any additional changes for those. Thanks!

omissis commented:
ofc, will do that soon

omissis committed:
Reintroduce support for php 5.6.

omissis committed:
Fix code style issue.

omissis committed:
Update the tests to be compatible with phpunit6.

omissis committed:
Fix code style.

omissis committed:
fix coding style issues.

omissis commented:
@padraic I've rebased on top of master and fixed all the style issues pointed out by styleCI, but I am not sure I'd still call this mergeable. I noticed Mockery 0.9 was introduced as a dev dependency to create some mocks. From what I can see Mockery depends on phpunit 5, which could potentially bring to conflicts or unexpected behaviours. I'd suggest -especially considering the limited size of this lib- to use phpunit's mocks directly and get rid of that dependency to keep things simpler.

I will pick up working on this one tomorrow.

padraic commented:
Mockery is perfect. I wrote it :).

On a serious note, you're right at the moment that it requires PHPUnit 5. I'll actually be upping that to 6 if possible separately. AFAIK, it should be compatible once I convert tests over to namespaced class names.

omissis commented:
eheh I know I know :P
However it seems it's not possible to support phpunit5 even through its FC layer, as `PHPUnit\Framework\Test` and `PHPUnit\Framework\TestSuite` are not supported.
My suggestion to close this sooner than later is to release a new phpunit6-only, mockery-free major. This should unlock all the deps that are currently blocking humbug, my real target :)

omissis committed:
implement test using only phpunit.

omissis committed:
fix cs issues.

omissis committed:
add a better explaination of a method implementation.

omissis committed:
remove php5 support.

omissis commented:
Hi, I took the time to review the new tests that were introduced in #5 and move them to phpunit 6.
I believe this highlights and resolve a few issue, such as:

**Testing a mock**
https://github.com/humbug/phpunit-extensions/blob/master/tests/Humbug/Test/Phpunit/Logger/JsonLoggerTest.php#L28

Even if partial, the SUT should not be mocked in order not to modify its behaviour

**Hidden dependecy**
https://github.com/humbug/phpunit-extensions/blob/master/src/Humbug/Phpunit/Logger/JsonLogger.php#L52

The need for partial mocks hints at hidden dependencies(violating SRP). In order to fix this, I implemented a JsonWriter to extract the logic and passed that as collaborator.

**Mocking not-existing methods**
https://github.com/humbug/phpunit-extensions/blob/master/tests/Humbug/Test/Phpunit/Listener/TimeCollectorListenerTest.php#L37

TimeCollectorListenerTest mocks getName() on Test, but that method does not exist. Phpunit (rightfully) complained the method did not exist on the interface. The problem is that the TimeCollectorListener::endTest() tries to use the getName() method which is not available on the hinted mock. I think it's more of a phpunit's shortcoming, but at the same time I don't think creating objects that do not exist (Test mock with getName()) is the best way to go.

Also I am not sure that relying on __destruct() to flush the log contents is the most reliable option, but I haven't looked into other options for now, but maybe some kind of callback would be easier to control (and test).

Thanks for taking the time to review this! :)

padraic commented:
The __destruct() use might be a shortcoming of using a Listener, or I might have simply "made it work" and never revisited it. I'd raise an issue separately to investigate a better alternative.

padraic reviewed (approved):
Merge only when 1.0 moved to maintenance branch.

theofidry commented:
Thanks @omissis for that hard work, gonna get PHPUnit6 on Humbug ASAP :)
